### PR TITLE
[Placement Group] Support named placement group java api & Refactor construct method

### PR DESCRIPTION
**Pull request by clay4megtr**
## Why are these changes needed?

This pr contains the following changes:
1. Refactor creation construct method of placement group and make it creation from a Option, the advantage of this is that if we want to add new features in the future, we no longer need to modify the constructor and jni interface, only need to modify the Option.
2. Implement the named placement group for java api

## Related issue number
#13625 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by clay4444** [comment]
Please do not merge this before this pr is merged:  [support named placement group in python](https://github.com/ray-project/ray/pull/13755)

File: java/api/src/main/java/io/ray/api/Ray.java
Comment (kfstorm) [inline java/api/src/main/java/io/ray/api/Ray.java:267]: API changes should be backward-compatible.

File: java/runtime/src/main/java/io/ray/runtime/task/NativeTaskSubmitter.java
Comment (kfstorm) [inline java/runtime/src/main/java/io/ray/runtime/task/NativeTaskSubmitter.java:101]: Shouldn't GCS check this rather than driver or worker?

**Review by kfstorm — COMMENTED** [review]
cc @pcmoritz for API changes.

File: java/runtime/src/main/java/io/ray/runtime/task/NativeTaskSubmitter.java
Comment (clay4444) [reply java/runtime/src/main/java/io/ray/runtime/task/NativeTaskSubmitter.java:101]: Currently, the placement group registration is async, the core worker will return `Status::OK` even If the actual registration is failed. but that's a good point because we already have a plan to make the registration sync.

File: java/api/src/main/java/io/ray/api/Ray.java
Comment (clay4444) [reply java/api/src/main/java/io/ray/api/Ray.java:267]: Updated.

**Review by kfstorm — APPROVED** [review]

File: java/api/src/main/java/io/ray/api/Ray.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/Ray.java:298]: gcs actor manager is always turned on now.

File: java/api/src/main/java/io/ray/api/Ray.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/Ray.java:346]: When you create a PG, do you specify that it is global?

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:55]: Check this.name is null?

File: java/api/src/main/java/io/ray/api/runtime/RayRuntime.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/runtime/RayRuntime.java:215]:
```suggestion
   * @param name The name of the placement group.
```

File: java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java
Comment (ffbin) [inline java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java:180]: Judge whether strategy is null when creating creationOptions?

File: java/runtime/src/main/java/io/ray/runtime/placementgroup/PlacementGroupUtils.java
Comment (ffbin) [inline java/runtime/src/main/java/io/ray/runtime/placementgroup/PlacementGroupUtils.java:97]: Why change this?

File: src/ray/core_worker/lib/java/io_ray_runtime_task_NativeTaskSubmitter.cc
Comment (ffbin) [inline src/ray/core_worker/lib/java/io_ray_runtime_task_NativeTaskSubmitter.cc:195]: RAY_CHECK java_name is not null?

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (clay4444) [reply java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:55]: yeah, that's a good idea.

File: java/runtime/src/main/java/io/ray/runtime/placementgroup/PlacementGroupUtils.java
Comment (clay4444) [reply java/runtime/src/main/java/io/ray/runtime/placementgroup/PlacementGroupUtils.java:97]: because the user can judge whether it exists with the name before create it.

File: src/ray/core_worker/lib/java/io_ray_runtime_task_NativeTaskSubmitter.cc
Comment (clay4444) [reply src/ray/core_worker/lib/java/io_ray_runtime_task_NativeTaskSubmitter.cc:195]: em... the default java_name is null

File: java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java
Comment (clay4444) [reply java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java:180]: yeah, I will move the bundles check to creationOptions too.

File: java/api/pom.xml
Comment (chaokunyang) [inline java/api/pom.xml:45]: add dependency inomp_template.xml and bazel BUILD file

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (kfstorm) [reply java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:68]: About the API naming, I think using `setName` with overloads is a better design. e.g.

```
public Builder setName(String name, boolean global) {
	this.name = name;
	this.global = global;
}

public Builder setName(String name) {
	return setName(name, /*global=*/false);
}
```

Comment (clay4444) [reply java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:68]: yeah, that's a good idea @kfstorm

File: java/api/pom.xml
Comment (clay4444) [reply java/api/pom.xml:45]: Yeah, thanks for reminding

File: java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java
Comment (chaokunyang) [inline java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java:185]: Should be `getPlacementGroupInfo`? `AbstractRayRuntime#getPlacementGroup` just return whatever `gcsClient.getPlacementGroupInfo` returns.

File: java/runtime/src/main/java/io/ray/runtime/gcs/GlobalStateAccessor.java
Comment (chaokunyang) [inline java/runtime/src/main/java/io/ray/runtime/gcs/GlobalStateAccessor.java:86]: Overloded methods should all have same return type. Would it be better to name it as `getPlacementGroup`?

File: java/test/src/main/java/io/ray/test/PlacementGroupTest.java
Comment (chaokunyang) [inline java/test/src/main/java/io/ray/test/PlacementGroupTest.java:45]: Will it be more clear to call like `setPlacementGroup(placementGroup).setPlacementGroupIndex(0)`?

Comment (clay4444) [reply java/test/src/main/java/io/ray/test/PlacementGroupTest.java:45]: That's maybe a better design, but actually, I don't think we should involve this core API change in this pr

File: java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java
Comment (clay4444) [reply java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java:185]: hm... I don't think so, the naming of `AbstractRayRuntime#getPlacementGroup` actually just follows the previous design like `AbstractRayRuntime#createActor` `AbstractRayRuntime#getActorHandle`

File: java/runtime/src/main/java/io/ray/runtime/gcs/GlobalStateAccessor.java
Comment (clay4444) [reply java/runtime/src/main/java/io/ray/runtime/gcs/GlobalStateAccessor.java:86]: sorry, I didn't get what's your mean, It seems that these methods are all return `byte[]`, right?  and the naming of `getxxxInfo` also just follow the previous design like `getNodeResourceInfo`、 `getAllNodeInfo` and `getAllJobInfo`

**Comment by kfstorm** [comment]
@ffbin @chaokunyang Any more comments?

File: java/runtime/src/main/java/io/ray/runtime/gcs/GlobalStateAccessor.java
Comment (chaokunyang) [reply java/runtime/src/main/java/io/ray/runtime/gcs/GlobalStateAccessor.java:86]: My fault, I took `GlobalStateAccessor#getPlacementGroupInfo` as `GcsClient#getPlacementGroupInfo` by mistake

**Review by chaokunyang — APPROVED** [review]

File: java/api/src/main/java/io/ray/api/Ray.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/Ray.java:331]:
```suggestion
   * @param name The placement group name.
```

File: java/api/src/main/java/io/ray/api/Ray.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/Ray.java:341]:
```suggestion
   * @param name The placement group name.
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:19]:
```suggestion
        bundles, "`Bundles` must be specified when creating a new placement group.");
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:29]:
```suggestion
        strategy, "`PlacementStrategy` must be specified when creating a new placement group.");
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:20]: Does the python API also do this check? Why not check it in the C++ code?

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:18]:
```suggestion
    Preconditions. checkArgument(bundles != null && !bundles.isEmpty()
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:45]:
```suggestion
     * this job by this name via {@link Ray#getPlacementGroup(java.lang.String)}. If you want to create
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:60]:
```suggestion
     * Set the name of a named placement group. This placement group can be accessed by all jobs via this name
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:77]:
```suggestion
     * resources on the raylet side.
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:91]:
```suggestion
     * @param strategy The placement strategy between bundles.
```

File: java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java
Comment (ffbin) [inline java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java:169]:
```suggestion
        "`PlacementGroupCreationOptions` must be specified when creating a new placement group.");
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (clay4444) [reply java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:20]: yeah, Currently, Python API will do the same check, maybe we can move it to the core worker in another independent pr if needed?

File: java/test/src/main/java/io/ray/test/PlacementGroupTest.java
Comment (ffbin) [inline java/test/src/main/java/io/ray/test/PlacementGroupTest.java:187]: No verification?

File: src/ray/core_worker/lib/java/jni_utils.h
Comment (ffbin) [inline src/ray/core_worker/lib/java/jni_utils.h:175]: strategyValue method of PlacementGroupCreationOptions class？

File: java/test/src/main/java/io/ray/test/PlacementGroupTest.java
Comment (clay4444) [reply java/test/src/main/java/io/ray/test/PlacementGroupTest.java:187]: Yes, because this is an exceptional case and the second creation will be failed.

File: src/ray/core_worker/lib/java/jni_utils.h
Comment (clay4444) [reply src/ray/core_worker/lib/java/jni_utils.h]: Yeah, that's a good point and I have changed it to `value method of PlacementStrategy class` just follow the previous naming

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (ffbin) [reply java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:21]: SGTM

**Review by ffbin — APPROVED** [review]

File: java/BUILD.bazel
Comment (kfstorm) [inline java/BUILD.bazel:42]: Please remove this large dependency from api module. API dependencies should keep as simple as possible.

**Review by kfstorm — APPROVED** [review]

File: java/BUILD.bazel
Comment (clay4444) [reply java/BUILD.bazel]: Removed!

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (kfstorm) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:57]:
```suggestion
        throw new IllegalArgumentException("Repeated assignment of name is not allowed!");
```

File: java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java
Comment (kfstorm) [inline java/api/src/main/java/io/ray/api/options/PlacementGroupCreationOptions.java:75]:
```suggestion
        throw new IllegalArgumentException("Repeated assignment of name is not allowed!");
```

**Pull request merged**
